File: algorithms/medium/xor_queries_of_a_subarray.ex
defmodule Solution do
  use Bitwise
  @spec xor_queries(arr :: [integer], queries :: [[integer]]) :: [integer]
  def xor_queries(arr, queries) do
    prefix =
      arr
      |> Enum.with_index()
      |> Enum.reduce(%{0 => 0}, fn {n, i}, acc ->
        Map.put(acc, i + 1, bxor(acc[i], n))
      end)

    queries
    |> Enum.map(fn [l, r] ->
      bxor(prefix[l], prefix[r + 1])
    end)
  end
end

File: algorithms/medium/xor_queries_of_a_subarray.rs
impl Solution {
    pub fn xor_queries(arr: Vec<i32>, queries: Vec<Vec<i32>>) -> Vec<i32> {
        let mut prefix = vec![0];
        for (i, &n) in arr.iter().enumerate() {
            let p = prefix[i] ^ n;
            prefix.push(p);
        }
        let mut ans = Vec::new();
        for q in &queries {
            let l = q[0] as usize;
            let r = q[1] as usize;
            ans.push(prefix[l] ^ prefix[r + 1]);
        }
        ans
    }
}
